lang/c++/impl/ValidSchema.cc

```diff
@@ -113,18 +113,83 @@ ValidSchema::setSchema(const Schema &schema)
     validate(root_);
 }
 
-void 
+void
 ValidSchema::toJson(std::ostream &os) const
-{ 
+{
     root_->printJson(os, 0);
     os << '\n';
 }
 
-void 
+std::string
+ValidSchema::toJson(bool prettyPrint) const
+{
+    ostringstream oss;
+    toJson(oss);
+    if (prettyPrint) {
+        return oss.str();
+    }
+
+    // Compact the formatted schema.
+    std::string schema = oss.str();
+    compactSchema(&schema);
+    return schema;
+}
+
+void
 ValidSchema::toFlatList(std::ostream &os) const
-{ 
+{
     root_->printBasicInfo(os);
 }
 
+/*
+ * compactSchema compacts a formatted string representation of a ValidSchema
+ * object by removing the whitespaces outside of the quoted field names and
+ * values. The schema string is modified in-place. It can handle the cases where
+ * the quoted value is in UTF-8 format. Note that this method is not responsible
+ * for validating the schema.
+ */
+void ValidSchema::compactSchema(std::string* schema) {
```

Comment:
The code assumes that `schema` is never `NULL`. So it is better to pass a reference instead of a pointer. This would prevent the caller from ever passing a `NULL`.
Having said that, I don't think it will be huge performance problem if we just return a new string instead of compacting in-place. That way one does not have to justify the `const_cast` here.